Cognito client crendetials flow (#1528)
localstack/docs: content/en/user-guide/aws/cognito/index.md

{"access_token": "eyJ0eXAi…lKaHx44Q", "expires_in": 86400, "token_type": "Bearer", "refresh_token": "e3f08304", "id_token": "eyJ0eXAi…ADTXv5mA"}
```

### Client credentials grant

The client credentials grant is designed for machine-to-machine (M2M) communication.
In contrast, the authorization code and implicit grants provide tokens to authenticated human users.
The client credentials grant allows for scope-based authorization from a non-interactive system to an API.
Your app can directly request client credentials from the token endpoint to receive an access token.

> To request the token from LocalStack the correct URL is `http://cognito-idp.localhost.localstack-test.cloud:4566/_aws/cognito-idp/oauth2/token`.

Reviewer: ```suggestion
To request the token from LocalStack the correct URL is `http://cognito-idp.localhost.localstack.cloud:4566/_aws/cognito-idp/oauth2/token`.
```

I think that you copy-pasted this from my issue, but the `-test` part was relevant for me because I had 2 LocalStack containers:
- `localstack`: for manual testing
- `localstack-test`: for automatic tests

Also, I have to say that the URL that worked for me was the following one:
```python
token_url = "http://localhost.localstack-test.cloud:4566/_aws/cognito-idp/oauth2/token"
```

However, I also had the domain configured for the user pool:
```sh
awslocal cognito-idp create-user-pool-domain \
  --domain "${COGNITO_USER_POOL_DOMAIN_PREFIX}" \
  --user-pool-id "${COGNITO_USER_POOL_ID}" \
  --region "${COGNITO_AUTH_REGION}"
```

Yet, supposedly, when a Domain is configured, the token should be accesible from another endpoint:
```
"https://mydomain.auth.us-east-1.amazoncognito.com/_aws/cognito-idp/oauth2/token"
```
(Doubt: in AWS, if the domain is configured are endpoints available or just the domain one?)

Based on: https://docs.aws.amazon.com/cognito/latest/developerguide/token-endpoint.html#post-token-positive 
> The token endpoint becomes publicly available when you add a domain to your user pool. It accepts HTTP POST requests. For application security, use PKCE with your authorization code sign-in events. PKCE verifies that the user passing an authorization code is that same user who authenticated. For more information about PKCE, see [IETF RFC 7636](https://datatracker.ietf.org/doc/html/rfc7636).

But I don't know if LocalStack is capable of replicating this behavior (due to being local and requiring HTTPS).

In case that there is more than one user pool, LocalStack detects the right one by inspecting the `clientId` of the requests.

Here is an example on how to set it up:

```sh
#Create user pool.
export pool_id=$(awslocal cognito-idp create-user-pool --pool-name test   --username-configuration "CaseSensitive=False"  | jq -rc ".UserPool.Id")

#Create client user pool with a client.
export client_id=$(awslocal cognito-idp create-user-pool-client --user-pool-id $pool_id --client-name test-client --generate-secret  | jq -rc ".UserPoolClient.ClientId")

#Retrieve secret.

> export client_secret=$(awslocal cognito-idp describe-user-pool-client --user-pool-id $pool_id --client-id $client_id | jq -r '.UserPoolClient.ClientSecret')

Reviewer: Take into account that if the LocalStack user has setup LocalStack as a separate container to their system, they won't be able to get the variables exported here.

What we did was have a local.env file in which the variables were defined (e.g. user pool id). Then pass that to the LocalStack container in the Environment section. Then, in the app use that same local.env to have the same environment information in both containers.

When creating the pool and so on, it would be cool to remind people that you can setup the custom_id, instead of letting LocalStack generate it.

#Create resource server
awslocal cognito-idp create-resource-server \
  --user-pool-id $pool_id \
  --identifier "api-client-organizations" \

>   --name "vitalera API Clients Organizations Resource Server" \

Reviewer: ```suggestion
  --name "Resource Server Name" \
```
We appreciate the ad here, and happy to appear, but maybe it's more correct if you put something generic here.

  --scopes '[{"ScopeName":"read","ScopeDescription":"Read access to Organizations"}]'

```

Then you could retrieve the token from your application like this:

```python
import requests, os
from requests.auth import HTTPBasicAuth

def get_access_token_with_secret():
    client_id = os.environ['client_id']
    client_secret = os.environ['client_secret']
    scope = 'api-client-organizations/read'
    url = 'http://cognito-idp.localhost.localstack.cloud:4566/_aws/cognito-idp/oauth2/token'
    

    headers = {
        'Content-Type' : 'application/x-www-form-urlencoded'
    }

    auth = HTTPBasicAuth(client_id,client_secret)

    payload = {
        'grant_type': 'client_credentials',
        'client_id':client_id,
        'scope':scope
    }
    response = requests.post(url,headers=headers,auth=auth,data=payload)

    print(response.content)

if __name__ == "__main__":
    get_access_token_with_secret()
```

> 

Reviewer: I think that it would be better to have a sample code in which the Cognito client is created in the Python code, from which you could extract the client ID and secret, and then do the token request. The sh would only need to create the pool.

I say this because that may be a more common scenario, normally you would create as many apps as M2M users are of your system.

{"access_token": "eyJ0eXAi…lKaHx44Q", "expires_in": 86400, "token_type": "Bearer", "refresh_token": "e3f08304", "id_token": "eyJ0eXAi…ADTXv5mA"}
```

### Client credentials grant

> The client credentials grant is designed for machine-to-machine (M2M) communication.
> In contrast, the authorization code and implicit grants provide tokens to authenticated human users.

Reviewer 2: ```suggestion
The client credentials grant is designed for machine-to-machine (M2M) communication.
The Client Credentials Grant allows the machine (client) to authenticate itself directly with the authorization server using its credentials, such as a client ID and client secret.
```

The client credentials grant allows for scope-based authorization from a non-interactive system to an API.
Your app can directly request client credentials from the token endpoint to receive an access token.

> To request the token from LocalStack the correct URL is `http://cognito-idp.localhost.localstack.cloud:4566/_aws/cognito-idp/oauth2/token`.

Reviewer 2: ```suggestion
To request the token from the LocalStack URL, use the following endpoint: `://cognito-idp.localhost.localstack.cloud:4566/_aws/cognito-idp/oauth2/token`.
```

> For additional information on our endpoints you can check our [Internal Endpoints](https://docs.localstack.cloud/references/internal-endpoints/) documentation.

Reviewer 2: ```suggestion
For additional information on our endpoints, refer to our [Internal Endpoints](https://docs.localstack.cloud/references/internal-endpoints/) documentation.
```

> In case that there is more than one user pool, LocalStack detects the right one by inspecting the `clientId` of the requests.

Reviewer 2: ```suggestion
If there are multiple user pools, LocalStack identifies the appropriate one by examining the `clientid` of the request.
```

> Here is an example on how to set it up:

Reviewer 2: ```suggestion
To get started, follow the example below:
```

```sh
#Create client user pool with a client.
export client_id=$(awslocal cognito-idp create-user-pool-client --user-pool-id $pool_id --client-name test-client --generate-secret  | jq -rc ".UserPoolClient.ClientId")

#Retrieve secret.
export client_secret=$(awslocal cognito-idp describe-user-pool-client --user-pool-id $pool_id --client-id $client_id | jq -r '.UserPoolClient.ClientSecret')

#Create resource server
awslocal cognito-idp create-resource-server \
  --user-pool-id $pool_id \
  --identifier "api-client-organizations" \
  --name "Resource Server Name" \
  --scopes '[{"ScopeName":"read","ScopeDescription":"Read access to Organizations"}]'

```

> Then you can retrieve the token from your application by using the mentioned endpoint: `http://cognito-idp.localhost.localstack.cloud:4566/_aws/cognito-idp/oauth2/token` endpoint.

Reviewer 2: ```suggestion
You can retrieve the token from your application using the specified endpoint: `http://cognito-idp.localhost.localstack.cloud:4566/_aws/cognito-idp/oauth2/token`.
```
